### PR TITLE
Serialized event handling

### PR DESCRIPTION
**Pull request by Lyude**
This is in response to the comments on #25 

Continuation of the work for serialized events. This also demonstrates how I was able to just combine the two futures we needed to run in parallel into one to make the interface simpler. Note that I've only tested this with my `neovim-gtk` fork so far, I'm going to try to make sure the nested requests example is still working hopefully this weekend and finish cleaning the rest of this up and implementing proper error handling.

Let me know what you think.

### CODE REVIEW

File: src/neovim.rs [hunk 1]
```diff
@@ -100,8 +106,11 @@ where
       queue: Arc::new(Mutex::new(Vec::new())),
     };
 
-    let req_t = req.clone();
-    let fut = Self::io_loop(handler, reader, req_t);
+    let (sender, receiver) = unbounded();
+    let fut = future::try_join(
+      Self::io_loop(reader, sender, req.clone()),
+      req.clone().handle_rec(handler, receiver)
+    ).map_ok(|_| ());
```
Comment (KillTheMule) [inline src/neovim.rs:113]: Why is this needed? Both futures return () on success, dont they?

File: examples/nested_requests.rs [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    match name.as_ref() {
+      "set_froodle" => {
+        *self.froodle.lock().await = args[0].as_str().unwrap().to_string()
+      }
+      _ => {}
+    };
+  }
+}
+
+#[tokio::main(basic_scheduler)]
```
Comment (KillTheMule) [inline examples/nested_requests.rs:109]: This needs updating, I think `(flavor = "current_thread")` should do it. We definitely want this one to run on just one thread (and analogously the test).

**Comment by KillTheMule** [comment]
So `try_join` seems to be able to reduce the futures the user needs to run from 2 to 1, good job! I've left two comments.

So, this is basically my branch `serial`'s last commit, apart from the `try_join`, isn't it? Maybe you could reinstate [this test](https://github.com/KillTheMule/neovim-lib/commit/62108d0e3e25383a10492eb5e2cc4adde41e4b65) (with single threaded runtimes, can we achive that with async-std as well?), just to make sure we're not missing anything.

**Comment by KillTheMule** [comment]
Oh, and we need to think about point 2 of [my earliest PR](https://github.com/daa84/neovim-lib/pull/23#issue-249877029) and see if that problem is solved by this design. Maybe we can find a test.

File: src/neovim.rs (hunk 1 above)
Comment (Lyude) [reply src/neovim.rs:113]: Exactly! Which means since `future::try_join()` returns the results from both futures, the actual return type is `((), ())` - which is a bit awkward, so we `.map_ok()` it back into a single `()`

**Comment by Lyude** [comment]
> So `try_join` seems to be able to reduce the futures the user needs to run from 2 to 1, good job! I've left two comments.
> 
> So, this is basically my branch `serial`'s last commit, apart from the `try_join`, isn't it? 
Maybe you could reinstate [this test](https://github.com/KillTheMule/neovim-lib/commit/62108d0e3e25383a10492eb5e2cc4adde41e4b65) (with single threaded runtimes, can we achive that with async-std as well?), just to make sure we're not missing anything.

Yeah - I've been cleaning it up a bit more locally: removing leftover comments, being a bit more careful with how we handle errors on the channels, etc. I was actually hoping to write some tests based off the examples as well though, but it looks like you've got some tests already though so I'd be happy to take a look at reinstating them to hopefully save myself some effort.
Another test I'm considering writing in addition to yours is one to actually start up a neovim `Child`, connect to it, then also connect to it with a second `Neovim<W>` instance that uses unix sockets and try interrupting sending/receiving on the second instance to make sure that we're propagating errors from the two joined futures correctly. More specifically: making sure that if one future ends pre-maturely because of a non-reader error and that results in the second future ending because of a reader error (which wouldn't be terribly useful to the user, since the reader error was just a result of the actual error from the first future that needs to be handled) that we forward the non-reader error to the user.

**Comment by Lyude** [comment]
> Another test I'm considering writing in addition to yours is one to actually start up a neovim `Child`, connect to it, then also connect to it with a second `Neovim<W>` instance that uses unix sockets and try interrupting sending/receiving on the second instance to make sure that we're propagating errors from the two joined futures correctly. More specifically: making sure that if one future ends pre-maturely because of a non-reader error and that results in the second future ending because of a reader error (which wouldn't be terribly useful to the user, since the reader error was just a result of the actual error from the first future that needs to be handled) that we forward the non-reader error to the user.

Actually - I'm being silly here and forgetting that only one future is handling I/O with nvim and the other is just handling the channel we created to pass along incoming events. So this would probably be a bit overkill

**Comment by Lyude** [comment]
> Actually - I'm being silly here and forgetting that only one future is handling I/O with nvim and the other is just handling the channel we created to pass along incoming events. So this would probably be a bit overkill

…and ignore this, whoops, because I realized we're still sending responses from handling requests to nvim in `handle_rec()` so we do still need to make sure errors get propogated correctly from both

**Comment by Lyude** [comment]
(next up on my TODO: see if we should still get the test you mentioned in https://github.com/KillTheMule/neovim-lib/commit/62108d0e3e25383a10492eb5e2cc4adde41e4b65 working)

**Comment by Lyude** [comment]
> So `try_join` seems to be able to reduce the futures the user needs to run from 2 to 1, good job! I've left two comments.
> 
> So, this is basically my branch `serial`'s last commit, apart from the `try_join`, isn't it? Maybe you could reinstate [this test](https://github.com/KillTheMule/neovim-lib/commit/62108d0e3e25383a10492eb5e2cc4adde41e4b65) (with single threaded runtimes, can we achive that with async-std as well?), just to make sure we're not missing anything.

So I looked into trying to get this test working again, but I honestly couldn't figure out what this test would accomplish that we aren't already doing with other tests (other than also just testing `ChildStdin`, which seems like it might be better for a separate MR). Am I missing something here?

Also - I think the rest of this branch is definitely ready for review

**Comment by KillTheMule** [comment]
Alright, awesome! I'll review, but I've also followed along and think this will fit nicely :) Please note I've just had surgery on my right hand, so everything will be pretty slow for some time. Especially, since there are 2 things I'll want to do before merging that (add a specific test that should have existed all along, and try out this PR with neovide), but I'm not sure when I'll have recovered enough for that.

**Review by KillTheMule — COMMENTED** [review]
Ok, I've made some comments, there basically nothing to do, although I've just "read" the code and could not yet try things locally. Hope I can get to that tomorrow :)

File: tests/notifications.rs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+     * notification executes pre-maturely if it was handled out of order.
+     */
+    yield_now().await;
+  }
+
+  // Check the results
+  for i in 0..ITERS {
+    assert_eq!(i as u8, timeout!(result_receiver.recv()).await.unwrap());
+  }
+}
```
[the remark this answers is not in the document]
Comment (Lyude) [reply tests/notifications.rs:125]: Sheesh-the one thing I forgot to do before submitting this :S, this test doesn't seem to actually work like I'd hope so I'll have to take another look at this

[the remark this answers is not in the document]
Comment (KillTheMule) [reply tests/notifications.rs:125]: The problem might be that it's not easy to make notification handling out-of-order even in the old implementation. It seemed only to be noticable on systems under load, which is kinda hard to emulate. Maybe add some task sleeping?

[the remark this answers is not in the document]
Comment (Lyude) [reply tests/notifications.rs:125]: Yeah-that seems to help. Something I realized as well though, which might let us still avoid using sleeps is- we don't technically know when nvim has finished running the commands we've passed to it. So it's entirely possible that by the time we've set off all of the notifiers, neovim hasn't even run anything in `-c` yet. probably could actually fix this by setting up a second rpc channel, and sending a single notification to it at the end so we know that neovim has finished sending us the notifications. I'll try that, but if that doesn't work I can just use some sleeps

[the remark this answers is not in the document]
Comment (Lyude) [reply tests/notifications.rs:127]: I ended up being able to fix this pretty nicely by just using `nvim.command(…)` for sending the rpc notifications instead of trying to pass them via `-c`. Since the callback doesn't return until neovim has finished executing the commands, that ensures that the notifications are in the queue.
Will continue working through the rest of the comments here :)

File: src/rpc/handler.rs [hunk 4]
```diff
@@ -31,7 +31,9 @@ pub trait Handler: Send + Sync + Clone + 'static {
     Err(Value::from("Not implemented"))
   }
 
-  /// Handling an rpc notification.
+  /// Handling an rpc notification. Notifications are handled one at a time in
+  /// the order in which they were received, and will block new requests from
+  /// being received until handle_notify returns.
```
[the remark this answers is not in the document]
Comment (Lyude) [reply src/rpc/handler.rs:36]: No problem!

File: src/neovim.rs [hunk 5]
```diff
@@ -180,39 +190,35 @@ where
...
     loop {
-      let msg = match model::decode(&mut reader, &mut rest).await {
-        Ok(msg) => msg,
-        Err(err) => {
-          let e = neovim.send_error_to_callers(&neovim.queue, *err).await?;
-          return Err(Box::new(LoopError::DecodeError(e, None)));
-        }
+      let msg = match receiver.next().await {
+        Some(msg) => msg,
+        None => break Ok(()),
```
[the remark this answers is not in the document]
Comment (Lyude) [reply src/neovim.rs:204]: `.unwrap()` sounds fine, I don't know if this error is actually worth propagating though. When I first was writing this I thought it might be back in https://github.com/KillTheMule/nvim-rs/pull/26#issuecomment-873438358 , but I realized I might be overthinking it - the only reason why `receiver` would close here is because `io_loop()` exited

[the remark this answers is not in the document]
Comment (Lyude) [reply src/neovim.rs:204]: Actually, I think `Ok(())` would be the right thing to do here? `receiver` is only receiving messages from `io_loop()` and not `handler_loop()`. And also, I would think that `receiver` is always going to close at some point anyway once `io_loop()` shuts down - properly or improperly. So I don't think we want to panic here.

[the remark this answers is not in the document]
Comment (KillTheMule) [reply src/neovim.rs:204]: Ohhh right that would be the normal shutdown procedure - neovim closes stdin/stdout, the io-loop shuts down (and now you mention it, that should not be an error as well, but just `Ok(())`) and takes the receiver with it. `Ok(())` it is, thanks! Please add a little comment to the function for my future self ;)

File: src/neovim.rs [hunk 6]
```diff
@@ -225,37 +231,59 @@ where
                 },
               };
 
-            model::encode(neovim.writer, response)
+            model::encode(writer, response)
               .await
               .unwrap_or_else(|e| {
                 error!("Error sending response to request {}: '{}'", msgid, e);
```
[the remark this answers is not in the document]
Comment (Lyude) [reply src/neovim.rs:241]: Added to TODO.md btw

**Comment by Lyude** [comment]
Seems like I've gone through all of the review comments, so this should basically be ready for merge :)

**Comment by Kethku** [comment]
I'd like to test this in neovide. If I've been following this thread correctly, I ran into similar issues with an early version of neovide and have since been using a fork with a dumb serialization hack. I'm hopeful that this PR will also fix my issues

**Comment by KillTheMule** [comment]
> Seems like I've gone through all of the review comments, so this should basically be ready for merge :)

Very awesome, thanks a lot! Unfortunately, due to my current circumstances (needed to be hospitalized again :( ) I won't be able to do some laste testing and playing around too soon, so I hope you're ok with me delaying the merge a bit. I sincerely hope I can get it done within the next week!

I tried to run neovide with it, but there were tokio incompatibilities, and I can't deal with stuff efficiently here, so I did not investigate further. Will try to do that later, too, though :)

**Comment by Lyude** [comment]
> Very awesome, thanks a lot! Unfortunately, due to my current circumstances (needed to be hospitalized again :( ) I won't be able to do some laste testing and playing around too soon, so I hope you're ok with me delaying the merge a bit. I sincerely hope I can get it done within the next week!
> 
> I tried to run neovide with it, but there were tokio incompatibilities, and I can't deal with stuff efficiently here, so I did not investigate further. Will try to do that later, too, though :)

Oh no! I'm sorry to hear that. Health comes first, so I'm totally fine with waiting a bit. If there's anything I could help out with let me know

File: tests/notifications.rs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+  let mut nvim_cmds = Vec::<String>::with_capacity(ITERS);
+  for i in 0..ITERS {
+    nvim_cmds.push(format!("call rpcnotify(1, 'idx', {})", i));
+  }
+
+  /* Start sending the notifications. We use nvim.command() instead of passing
+   * the commands via -c on the command line so that we block the test until the
+   * notifications are actually pending.
+   */
+  timeout!(nvim.command(nvim_cmds.join("|").as_str())).await.unwrap();
```
Comment (KillTheMule) [inline tests/notifications.rs:109]: So I could look a bit more at the PR, and I'm kindof embarassed I don't understand this. I have this problem with almost all the tests, but maybe you've understood this better than I do right now...

So, `nvim.command` is a request, i.e. a future that resolves after we got the corresponding response. The response will be trivial, but it needs to be there for the future to resolve. But... where is it coming from? Responses are read by the `_io_handler` future, and you're not running that anywhere. I really feel daft (after all, the other tests do this as well and do pass), but... who's reading the response and delivering it to this future?

**Comment by KillTheMule** [comment]
Ok, I've written the test I've wanted and it works :) I couldn't push to your branch, so I'll just merge this afterwards. I hope tonight I can look at the neovide situation, and then I'll merge this.

**Comment by Kethku** [comment]
Don't block on neovide. I can test that after

**Comment by KillTheMule** [comment]
Well, I really want to see if it solves the problem neovide had with non-serialized messages. But locally, I can't reproduce said problem with current master, even when putting my system under heavy load (upgrading to tokio 1.1.1 was very easy). But neovide is very sluggish here anyways, so maybe it's sort of the graphics slowing it down so much that the notification pipeline appears serialized in effect? Could you try to reproduce that with current master? I can send you a PR for upgrading tokio.

**Comment by Kethku** [comment]
I will try to test this shortly

**Comment by Kethku** [comment]
I just tested both master and the serial fork and found issues in neither. That said, that sorta makes sense as I did some work recently to add a layer of serialization to neovim events on our side as well. This in combination with a refactoring a number of months ago which moves all of the processing of editor and renderer bits to their own threads means there is very little work happening in the actual event processing handler.

So I have closed the pr updating tokio on our side and created a new one which targets master nvim-rs. I will merge that pr once this PR has merged.

TLDR looks A-OK to me. Awesome stuff, very excited

**Comment by KillTheMule** [comment]
Awesome, thanks for checking! I'll make a new crates.io release soon.

Thanks for all the work and your peseverance, @Lyude !
